[Autocomplete] Clicking anywhere on Input should open listbox

hbjORbj opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Closes https://github.com/mui-org/material-ui/issues/29381

Problem:
- Only clicking `<input>` inside `InputBase` div opens the list box. It is generally expected that clicking anywhere on the `InputBase` div should open the listbox.
- preview: https://mui.com/components/autocomplete/

https://user-images.githubusercontent.com/758437/139541585-cdbf5b6b-f021-48bb-ab9c-0c75f2837f71.mp4

Solution:
- Remove `onMouseDown` event from  the nested `<input>` element 
- Attach `onClick` event to `InputBase` div
- make some small changes and use `event.stopPropagation()` properly on handlers to keep all other behaviours the same.
- preview: https://deploy-preview-29500--material-ui.netlify.app/components/autocomplete

https://user-images.githubusercontent.com/32841130/141288000-b681cfa6-fb7e-437e-a1fd-ddb28acd12e7.mov

hbjORbj commented on packages/mui-core/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -945,8 +945,19 @@ export default function useAutocomplete(props) {
+    // Since other handlers (e.g., handlePopupIndicator()) already exist to handle clicking specific elements inside the container,
+    // this handler performs the following conditional check to ensure that it runs only when <input /> or the `InputBase` div container is clicked
+    if (
+      typeof className === 'string' &&
```
I found that sometimes `typeof event.target.className` is `Element`.

hbjORbj commented on packages/mui-core/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -945,8 +945,19 @@ export default function useAutocomplete(props) {
+    // this handler performs the following conditional check to ensure that it runs only when <input /> or the `InputBase` div container is clicked
+    if (
+      typeof className === 'string' &&
+      (className.includes('InputBase-root') || className.includes('InputBase-input'))
```
I didn't add `Mui` because prefixes of classname are modifiable by the user now.

mnajdova commented on packages/mui-core/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -945,8 +945,19 @@ export default function useAutocomplete(props) {
   const handleInputMouseDown = (event) => {
-    if (inputValue === '' || !open) {
-      handlePopupIndicator(event);
+    const className = event.target.className;
```
Instead of all this, why can't we just add a `onMouseDown` on the `popupIndicator` that would stop propagating?

mnajdova replied:
Probably for the `cancelIndicator` too

hbjORbj commented on packages/mui-base/src/AutocompleteUnstyled/useAutocomplete.js:
```diff
@@ -1012,7 +1013,10 @@ export default function useAutocomplete(props) {
       onBlur: handleBlur,
       onFocus: handleFocus,
       onChange: handleInputChange,
-      onMouseDown: handleInputMouseDown,
```
I decided to use replace `onMouseDown` with `onClick` because this approach requires less change. 

For example, when `multiple` is enabled, we pass `onDelete: handleTagDelete()` to `Chip`s, which are used for multiple values. `Chip`'s `onDelete` API uses a click event, so dealing with event bubbling gets more complicated.

oliviertassinari commented on packages/mui-material/src/Autocomplete/Autocomplete.js:
```diff
@@ -519,6 +519,8 @@ const Autocomplete = React.forwardRef(function Autocomplete(inProps, ref) {
     });
   };
 
+  const { onClick: onInputClick, ...inputProps } = getInputProps();
```
This onClick prop fork seems to defeat the purpose of having `getInputProps()`.  All the props are supposed to land on the same DOM element.

oliviertassinari reviewed (commented):
It breaks this demo https://deploy-preview-29500--material-ui.netlify.app/components/autocomplete/#playground

<img width="339" alt="Screenshot 2021-11-11 at 17 31 10" src="https://user-images.githubusercontent.com/3165635/141334000-8f126f1d-72bd-4c0d-9c9d-71a5b5cfb288.png">

To be honest, I think that we could ignore the issue, in the first place (I recall doing this in the past). There are Autocomplete issues that are easier to solve with more upvotes, e.g. #18450.

hbjORbj commented:
@oliviertassinari Thanks for review. I addressed your comments.

> Also, I'm not sure about triggering the popup on mouseup. It was done on mouseup to reproduce the <select> on macOS, select2, and react-select behaviors.

To be fair, `handleClear` (clear icon) and `handlePopupIndicator` (popup icon) were already using a click event. So, there is only 1 change in terms of `event` (Input). Even though using `click` instead of `mouseDown` may not be ideal, I think fixing this bug is more important than the difference between the two events. What do you think?

hbjORbj commented:
@oliviertassinari

> Also, I'm not sure about triggering the popup on mouseup. It was done on mouseup to reproduce the `<select>` on macOS, select2, and react-select behaviors.

I agree with you on this point. It looks little laggy. I will explore other approaches.

siriwatknp commented:
> @oliviertassinari
> 
> > Also, I'm not sure about triggering the popup on mouseup. It was done on mouseup to reproduce the `<select>` on macOS, select2, and react-select behaviors.
> 
> I agree with you on this point. It looks little laggy. I will explore other approaches.

Have you figured it out?

mnajdova commented:
@hbjORbj what's the status here? Would be great to have the fix.

mnajdova commented:
> i think https://github.com/mui/material-ui/pull/36369 PR also solving same issue

Correct, let's close this one.

hbjORbj commented:
Yes, we can close it
